Complex numbers

scimind2460 opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3892)*

<!-- Thank you for trying to improve Rust through the RFC process! -->
<!-- Please add a short summary of your RFC below -->

This RFC proposes FFI-compatible complex numbers to help scientific computing library authors use non-indirected complexes.

I apologise in advance to `num-complex`


[Rendered](https://github.com/rust-lang/rfcs/blob/master/text/3892-complex-numbers.md)

scimind2460 committed:
Add complex numbers rfc

scimind2460 committed:
Change file name and pr to reflect current issue

joshtriplett commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,214 @@
+```
+But the easiest way to usually instantiate them is by using them like this:
+```rust
+let easy = 1 + 2.i()
````
I don't think the core RFC should be proposing a `.i()` method like this. There are multiple ways we could provide such syntax, including a method like `2.i()` method, a constant `I` to use like `2*I`, or native syntax like `2i` or `2j`.

Please move the mentions of `.i()` or similar shorthands to the "future possibilities" section as potential alternatives to native syntax.

kennytm commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,214 @@
+  fn computes_function(x: Complex<f64>) -> Complex<f64>;
+}
+fn main() {
+  let returned_value = computes_function(Complex<f64>::new(3, 4))
```
```suggestion
  let returned_value = computes_function(Complex::<f64>::new(3, 4))
```

clarfonthey commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,214 @@
+
+Complex numbers will be implemented by using public traits in the `core` crate that implement as an abstraction numeric features such as:
+```
+trait PrimNum: Add + Sub + Mul + Div + Neg + SimdElement + Copy + Clone {
````
As a good precedent: `NonZero` uses a sealed, unstable trait to allow `NonZero<T>` for a set of types without guaranteeing anything else about `T`.

clarfonthey commented:
It's worth pointing out another big issue with this is that the canonical `a+bi` is not actually the best representation of complex numbers in all cases, and so deciding on this is making a decision that might make life harder for external complex-numeric libraries out there.

In particular, while `a+bi` (orthogonal) representation is efficient for addition, `r*(iθ).exp()` is more efficient for multiplication, and depending on the equation you're using, it may be advantageous to switch between the two to reduce the number of arithmetic operations needed.

I'm not super compelled by the argument that C supports this, therefore the standard library needs to support this. I think that guaranteeing a `std::ffi::Complex` representation would be desirable, but there's no saying that we need to make this a canonical type in, say, `std::num`.

tgross35 commented:
> It's worth pointing out another big issue with this is that the canonical `a+bi` is not actually the best representation of complex numbers in all cases, and so deciding on this is making a decision that might make life harder for external complex-numeric libraries out there.
> 
> In particular, while `a+bi` (orthogonal) representation is efficient for addition, `r*(iθ).exp()` is more efficient for multiplication, and depending on the equation you're using, it may be advantageous to switch between the two to reduce the number of arithmetic operations needed.

I think that polar form almost always is the more optimal form, at least in my experience. But the ABIs do use rectangular, e.g. from x86:

> Arguments of complex T where T is one of the types float or double are treated as if they are implemented as:
> ```c
> struct complexT {
>   T real;
>   T imag;
> };
> ```

so it makes sense that an interchange type matches that, and users can translate to/from a polar repr at the FFI boundary if needed. But this reasoning is definitely something to have in the RFC's rationale & alternatives.

bluebear94 commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,214 @@
+If you want to call certain C libraries with complex numbers, you use this type:
+```C
+// in the C library
+extern double _Complex computes_function(x: double _Complex);
````
```suggestion
extern double _Complex computes_function(double _Complex x);
```

scimind2460 committed:
Incorporate feedback into RFC

scimind2460 committed:
Fix errors

tgross35 commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,332 @@
+}
+```
+to properly classify all types complex numbers can be implemented on.
+They will have an internal representation of a Tx2 array:
````
As mentioned earlier, the RFC doesn't need to specify internal details. It should actually avoid it here because we don't want to tell everyone that they can transmute this type to `[T; 2]`.

The actual repr is meaningless anyway, it could look like `[T; 2]`, `{ re: T, im: T }`, or just be empty—the compiler is going to do ABI magic that disregards what that syntax usually does.

SciMind2460 replied:
Thanks for informing me! Fixed that now.

scimind2460 committed:
Remove impls and replace with stubs

scimind2460 committed:
Fix polars

programmerjake commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,251 @@
+  fn computes_function(x: Complex<f64>) -> Complex<f64>;
+}
+fn main() {
+  let returned_value = computes_function(Complex::<f64>::new(3, 4))
```
```suggestion
  let returned_value = computes_function(Complex::<f64>::new(3.0, 4.0))
```

kennytm commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,251 @@
+
+`Complex<T>` numbers can be instantiated as of any type using `Complex::new(re, im)` where `re` and `im` are of the same type (this includes all numbers).
+```rust
+let x = Complex::new(3.0, 4.0); // this instantiates them as integers, not floats!
````
this comment does not make sense anymore.

```suggestion
let x = Complex::new(3.0, 4.0);
```

kennytm commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,251 @@
+  }
+  fn from_polar(modulus: f64, angle: f64) -> Complex<f64> {
+  }
+}
```
```suggestion
impl<T: Float> Complex<T> {
  fn angle(self) -> T;
  fn from_polar(modulus: T, angle: T) -> Self;
}
```

scimind2460 committed:
Add more future possibilities and fix typos

scimind2460 committed:
Add associated consts and collapse functions

scimind2460 committed:
Apply suggestions from code review

Co-authored-by: kennytm <kennytm@gmail.com>

SciMind2460 commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,251 @@
+  }
+  fn im(self) {
+  }
+}
```
Done now.

juntyr commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,233 @@
+
+impl Complex<f32> {
+  fn angle(self) -> f32;
+  fn from_polar(modulus: f32, angle: f32) -> Complex<f32> {
```
```suggestion
  fn from_polar(modulus: f32, angle: f32) -> Self;
```

juntyr commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,233 @@
+have construction methods and `From` impls:
+```rust
+impl Complex<T> {
+  fn new(re: T, im: T);
````
```suggestion
  fn new(re: T, im: T) -> Self;
```

scimind2460 committed:
Add fixes in type signature

Co-authored-by: Juniper Tyree <50025784+juntyr@users.noreply.github.com>

scimind2460 committed:
Apply suggestions from code review

Co-authored-by: Juniper Tyree <50025784+juntyr@users.noreply.github.com>

SciMind2460 commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,214 @@
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Some people on IRLO believe that complex numbers should be part of the std due to them being a language feature in many other languages including C, C++ and Go. They also mention that if Rust aims for good FFI with C, then complex numbers are an important feature.
```
Added all in the RFC, but moved the `.i()` to the `Future Possibilities` section since I don't want to commit to anything concrete. I could also move the polar proposal down there, but @clarfonthey doesn't seem too happy with the orthogonal representation, so I added that for easy convertibility.

jdahlstrom commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,245 @@
+```rust
+let polar = Complex::from_polar(3.0, f32::PI/2.0);
+```
+where .i() turns a real number into a complex one transposing the real value to a complex value.
````
The phrasing implies that `.i()` is used in the preceding example but it's not (this is the first mention that there's a thing called `.i()`)

scimind2460 committed:
Clarified some details regarding `i`

scimind2460 committed:
Fix wording in complex numbers documentation

SciMind2460 commented:
@clarfonthey one of the other reasons I don't agree with polar representation is that it makes it very hard to define complex integers (Gaussian integers) which I included as a future possibility and which are useful for dealing with discrete geometry.

SciMind2460 commented:
@clarfonthey IMO if you consider the fact that complex integers are only
really supported orthogonally, then it becomes clear that there is only
really one choice. Though polar representations are more accurate, they
simply wouldn't work with complex integers (Gaussian integers) which I
proposed as a future possibility. Rust has to decide in favor of the
orthogonal representation if we choose to support Gaussian integers.

On Fri, 5 Dec, 2025, 00:37 Clar Fon, ***@***.***> wrote:

> *clarfonthey* left a comment (rust-lang/rfcs#3892)
> <https://github.com/rust-lang/rfcs/pull/3892#issuecomment-3613134755>
>
> The conversion between a polar and orthogonal form isn't lossless, so, it
> effectively can't be done "automatically" or "as an implementation detail."
> You need trigonometric functions to do it, and while the conversion is
> algebraically closed, it's certainly more complicated to do exactly and
> most people prefer to just use floats instead.
>
> My point here isn't that we need to decide; the issue is that the decision
> itself means that we *shouldn't* decide, and instead avoid having a
> standard Complex type for the standard library.
>
> This doesn't preclude adding std::ffi::Complex which allows
> ABI-compatibility with C's _Complex, however, I don't think that such a
> type should be made standard for the language *because* of the fact that
> there are so many different ways to go about it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/3892#issuecomment-3613134755>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLL7PYEKKLORP5MEG4D3QK34ABPK3AVCNFSM6AAAAACNY66HSWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMJTGEZTINZVGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

programmerjake commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,243 @@
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The rationale for this type is mostly FFI: C libraries that may be linked from Rust code currently cannot provide functions with direct struct implementations of Complex - they must be hidden under at least a layer of indirection. However, it is not always possible to write a C complex-valued function that wraps the first function in a pointer. Thus, FFI becomes a problem if such complex-valued f ...
```
on `powerpc64-linux-gnu`, returning `double _Complex` doesn't do the same thing as returning a struct with a field of type `double[2]`: https://gcc.godbolt.org/z/hh7zYcnK6

programmerjake commented:
> 1. Provide a `Complex<T>` type which is FFI-compatible with C's `_Complex T` at least for `T = f32`, `f64`, and exposes [all methods available in C](https://en.cppreference.com/w/c/numeric/complex.html), or
> 2. Declare that [`[T; 2]` is FFI-compatible with `_Complex T`](https://github.com/rust-lang/rfcs/pull/3892#discussion_r2591195612) and be done with it. I think it is correct for Clang and GCC, but seems not the case for MSVC.

Deciding the in-memory representation is the easy part, the hard part is how exactly complex numbers are passed by value in function arguments and return values -- I recall reading ABI specs that treat complex numbers specially such that they don't really match the ABI of any other single type (so it has to be handled specially by rustc and can't just be an existing type), though I can't currently recall which ABI specs.

SciMind2460 commented:
@programmerjake you're right - the calling convention of complex numbers is really not defined very well, which is why in my opinion No. 2 seems untenable. Sometimes complex numbers are passed as a struct, sometimes SSE registers - it really is untenable for anything other than STD to match the calling convention.

scimind2460 committed:
Add possible workaround

Added theoretical C code example and alternatives for complex number handling in FFI.

scimind2460 committed:
Update rationale for Complex type in FFI context

Clarified rationale for Complex type regarding FFI and C calling conventions.

SciMind2460 commented:
No, what I was intending to say was that the fact that complex numbers
would be introduced would make some libraries use the std complexes
immediately whereas others would still use num-complex. I'll add
implementation complexity though.

On Sun, 7 Dec, 2025, 06:54 nora, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In text/3892-complex-numbers.md
> <https://github.com/rust-lang/rfcs/pull/3892#discussion_r2595642450>:
>
> > +impl Div for Complex<f32> { // calls to __divsc3 will be required here for implementation details and corresponding real types will also be implemented
> +  type Output = Self;
> +
> +  fn div(self, other: Self) -> Self::Output;
> +}
> +impl Div for Complex<f64> { // calls to __divdc3 will be required here for implementation details and corresponding real types will also be implemented
> +  type Output = Self;
> +
> +  fn div(self, other: Self) -> Self::Output;
> +}
> +```
> +The floating point numbers shall have sine and cosine and tangent functions, their inverses, their hyperbolic variants, and their inverses defined as per the C standard and with Infinity and Nan values defined as per the C standard.
> +## Drawbacks
> +[drawbacks]: #drawbacks
> +
> +If there is suddenly a standard-library Complex type, people may rush to include it in their current implementations, which would leave people behind if they didn't know about it. I really don't think this is a drawback though, since similar things have happened in Rust before: the inclusion of `OnceCell` in Rust, for example.
>
> I don't really understand what exactly this means. Who would want to rush
> to include complex numbers in their code? And who would be be left behind?
> I think the biggest downside here is the increased complexity and API
> surface and implementation maintenance.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/3892#pullrequestreview-3548517940>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLL7PYAVAPV6XHQUYQJ6OO34ANND5AVCNFSM6AAAAACNY66HSWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKNBYGUYTOOJUGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

programmerjake commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,252 @@
+## Drawbacks
+[drawbacks]: #drawbacks
+
+If there is suddenly a standard-library Complex type, people may rush to include it in their current implementations, which would leave people behind if they didn't know about it. I really don't think this is a drawback though, since similar things have happened in Rust before: the inclusion of `OnceCell` in Rust, for example.
```
github put [this reply](https://github.com/rust-lang/rfcs/pull/3892#issuecomment-3621349079) in the wrong thread, afaict it was intended to be here.

scimind2460 committed:
Update drawbacks section for Complex type proposal

Clarify potential drawbacks of a standard-library Complex type and its impact on implementation.

scimind2460 committed:
Improve clarity and rationale in complex numbers documentation

Clarified language regarding division and method descriptions for complex numbers. Expanded rationale for the complex number type and its implications for FFI.

joshtriplett commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,252 @@
+// in core::complex
+#[lang = "complex"] // For matching the calling convention (special repr needed?)
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct Complex<T: Float>(re: T, im: T);
```
```suggestion
pub struct Complex<T>(re: T, im: T);
```

programmerjake commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,252 @@
+```
+that could help simplify the life of people who otherwise would have to keep writing `Complex::new()`?
+- Should we support Imaginary eventually? This RFC doesn't cover it, but I think we can do this later in another RFC.
+- Eventually we may support Gaussian integers (an extension of the real integers) which have a Euclidean division procedure with remainder. We could theoretically eventually support these integers?
````
I'll note GCC has integer complex numbers as a C extension, which I think Rust should eventually support FFI with.

scimind2460 committed:
Apply suggestions from code review

Co-authored-by: Josh Triplett <josh@joshtriplett.org>

ComputerDruid commented:
If the motivation for putting it in std is just for FFI, could the std one just be a struct with pub fields and no otger API, and then have the nice complex type be provided by a crate which is a `#[repr(transparent)]` wrapper around that?

SciMind2460 commented:
> If the motivation for putting it in std is just for FFI, could the std one just be a struct with pub fields and no otger API, and then have the nice complex type be provided by a crate which is a `#[repr(transparent)]` wrapper around that?

Another problem that this RFC aims to resolve, which @miikkas pointed out, is to unify the API of complex numbers. Providing the std type as just a complex number with no methods would go against that.

scimind2460 committed:
Update RFC on complex numbers support

Clarify future support for Gaussian integers and floating point types in the RFC.

kpreid commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,252 @@
+impl Float for f32 {}
+impl Float for f64 {}
+```
+Calls to some `libgcc` functions will also be needed and will be emitted by the backend via compiler-builtins, specifically `__mulsc3`, `__muldc3`, `__divsc3` and `__divdc3` for the proper and complete implementation of these types.
````
Presumably these impls will be in `std`, and `core` will have the `Complex` type but without a full set of arithmetic impls, so `Complex` can still be used without dependency on system libraries. I think that might be worth describing more precisely and explicitly, for readers concerned with `no_std` use cases.

kpreid commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,252 @@
+### Alternatives:
+- Don't do this: There are, obviously, millions of alternatives on crates.io, the foremost being `num-complex`. However, I believe that if we wish to support proper FFI with C, then a standard type that matches calling conventions with C complex numbers is an important feature of the language. Hence, I do not recommend this idea.
+- Use a polar layout: Polar complex numbers, are undoubtedly a more optimal solution for multiplying complexes. However, I believe that if we wish to have proper FFI with C, then complex number layout should be chosen in accordance with the layout that is used in the C standard, and that is the orthogonal layout. This is also the layout used by most of other languages and crates on crates.io.
+- Non-generic primitive types: These are, obviously, the most obvious and practical solution. However, if we implemented lots of such types, then we would not be able to expand for `f16` and `f128` support without repeating the code already implemented. It would be extremely repetitive and tedious to add new types, especially since Gaussian integers and other floating points could have added supp ...
```
“Extremely repetitive and tedious” isn’t an *implementation* concern, because the standard library already [extensively uses macros](https://doc.rust-lang.org/1.91.0/src/core/num/mod.rs.html#295-314) for numeric impls. However, such implementations do make for terrible documentation (see, e.g. the current [NonZero](https://doc.rust-lang.org/1.91.0/std/num/struct.NonZero.html) doc page), which is a point in favor of generic implementations.

A point *against* generic implementations is that they currently cannot be `const fn`s. But hopefully const trait impls will arrive soon enough to make that not too painful.

scimind2460 committed:
Apply suggestions from code review

Co-authored-by: Kevin Reid <kpreid@switchb.org>

SciMind2460 commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,252 @@
+```c
+double _Complex function(double _Complex value);
+double _Complex *wrapper_function(double _Complex* value) {
+  return &function(*value);
````
Yep, that was a mistake on my part.

scimind2460 committed:
Enhance rationale and examples for complex numbers API

Clarified the rationale for a unified API for complex numbers and addressed compatibility issues with C libraries. Updated examples and alternatives to improve understanding of complex number implementations.

scimind2460 committed:
Add suggestion

Co-authored-by: Kevin Reid <kpreid@switchb.org>

scimind2460 committed:
Remove `this`

Amanieu commented:
I would like @cuviper (as the author of num-complex) to review the proposed API here.

scimind2460 committed:
Refactor Complex struct and update methods

Refactor Complex struct to have public fields and update methods for real and imaginary parts. Added notes on future considerations for polar conversions and arithmetic implementations.

scimind2460 committed:
Apply suggestions from code review

Co-authored-by: Josh Triplett <josh@joshtriplett.org>

SciMind2460 commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,252 @@
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The rationale for this type is mostly FFI: C libraries that may be linked from Rust code currently cannot provide functions with direct struct implementations of Complex - they must be hidden under at least a layer of indirection. This is because of the undefined calling convention of complex numbers in C. For example: on powerpc64-linux-gnu, [returning double _Complex doesn't do the same thing a ...
+Additionally, another issue this solves is to have a unified API for complex numbers. Right now, many crates are using their own implementation (`num-complex` could serve as a unifying factor, but other crates do not implement the same complex numbers like `nalgebra`), which makes it difficult to implement unifying interfaces without complicating the code with too many conversion functions. This  ...
```
1. Mentioned in the RFC -- they don't really care for compatibility.
2. Yes, they would, I suppose

cuviper commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,227 @@
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The rationale for this type is mostly FFI: C libraries that may be linked from Rust code currently cannot provide functions with direct struct implementations of Complex - they must be hidden under at least a layer of indirection. This is because of the undefined calling convention of complex numbers in C. For example: on powerpc64-linux-gnu, [returning double _Complex doesn't do the same thing a ...
+Additionally, another issue this solves is to have a unified API for complex numbers. Right now, many crates are using their own implementation (`num-complex` could serve as a unifying factor, but other crates do not implement the same complex numbers, such as `nalgebra`, due to less care over incompatibility concerns), which makes it difficult to implement unifying interfaces without complicatin ...
```
You're mistaken about `nalgebra`, at least -- it _re-exports_ `num_complex::Complex` in its own API so you don't have to depend on `num-complex` yourself, but it is still the same unified type.

https://docs.rs/nalgebra/latest/src/nalgebra/lib.rs.html#169

https://github.com/dimforge/nalgebra/blob/96c5d876f49afbcb0c22748b4f548b654cfef4fe/src/lib.rs#L169

SciMind2460 replied:
It appears I made a mistake, because [these docs](https://docs.rs/nalgebra/latest/nalgebra/struct.Complex.html) link back to nalgebra::Complex  as its own struct, but it is actually there in the definition of `num-complex`. I'll rework the example.

scimind2460 replied:
Resolved

cuviper commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,227 @@
+```
+that could help simplify the life of people who otherwise would have to keep writing `Complex::new()`?
+- Should we support Imaginary eventually? This RFC doesn't cover it, but I think we can do this later in another RFC.
+- Eventually we may support Gaussian integers (an extension of the real integers) which have a Euclidean division procedure with remainder. GCC has these, and we could theoretically eventually support these integers alongside GCC FFI.
````
FWIW, a recent issue raised that `num-complex`'s remainder is not ideal for their purpose: https://github.com/rust-num/num/issues/445

Although as I commented there, it can be seen similarly to the `Rem` vs. `rem_euclid` methods for integers in the standard library.

cuviper reviewed (commented):
> I apologise in advance to `num-complex`

No worries! There are definitely things that a compiler-integrated implementation can do better, especially with ABI as you've highlighted. A lot of the `num-complex` API is also limited by trying to be so generic, whereas the standard library can choose to keep all the implementation flexibility for itself.

scimind2460 committed:
Clarify reason for intrinsic calls

joshtriplett committed:
Apply suggestion from @programmerjake

Co-authored-by: Jacob Lifshay <programmerjake@gmail.com>

joshtriplett commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,225 @@
+They can even be passed in polar form (but only as a float):
+```rust
+let polar = Complex::from_polar(3.0, f32::PI/2.0);
+```
````
```suggestion
```

joshtriplett committed:
Remove polar from the main body of the RFC

tgross35 commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,221 @@
+The definition of complex numbers in the C99 standard defines the _memory layout_ of a complex number but not its _calling convention_. 
+This makes crates like `num-complex` untenable for calling C FFI functions containing complex numbers without at least a level of indirection (`*const Complex`) or the like.
+Only in `std` is it possible to make an additional repr to match the calling convention that C uses across FFI boundaries. 
+In essence, this RFC makes code like this:
```
```suggestion
The C standard defines the _memory layout_ of a complex number, but not their _calling convention_. 
This means crates like `num-complex` require workarounds to interface with FFI using `_Complex`, and cannot pass values directly.

In essence, this RFC makes code like this:
```

Readability cleanup

tgross35 commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,221 @@
+```rust
+let val = Complex::new(3.0, 4.0);
+let sine_cmplx = csin(val); // 3.8537380379 - 27.016813258i
+```
````
Delete this, mention trig and further functions in future possibilities instead. We wouldn't have a freestanding `csin` function anyway.

tgross35 commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,221 @@
+fn main() {
+  let returned_value = computes_function(Complex::<f64>::new(3.0, 4.0))
+}
+```
````
```suggestion
This type is FFI-compatible with `_Complex` in C for floating point types.
```

The same example is already repeated above.

tgross35 commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,221 @@
+[drawbacks]: #drawbacks
+
+The implementation surface of the complex types means more items for the libs and lang teams to maintain.
+Also, the multiple emitted calls to `libgcc.so` (`__mulsc3` and the like) via compiler-builtins may cause a bit of overhead and may not be what the Rust lang team and compiler team want.
```
This drawback doesn't exactly make sense, there aren't really any alternatives in many cases.

tgross35 commented on text/3892-complex-numbers.md:
````diff
@@ -0,0 +1,221 @@
+Calls to some `libgcc` functions may also be needed, and will be emitted by the backend via compiler-builtins, specifically `__mulsc3`, `__muldc3`, `__divsc3` and `__divdc3` for the proper and complete implementation of these types. This is because if we implemented these operations generically, to implement any overflow checks, we would need to implement a trait that can generically call, say, a ...
+They will have an internal representation similar to this (with public fields for real and imaginary parts):
+```rust
+// in core::complex
````
I think it's worth a section in the guide-level explanation saying that the type can be imported from `core::complex`. Since location has been a discussion point, and to make it obvious this isn't a prelude type.

scimind2460 replied:
Changed it to core::num::complex

scimind2460 committed:
Apply suggestions from code review

Co-authored-by: Trevor Gross <tg@trevorgross.com>

scimind2460 committed:
Update 3892-complex-numbers.md

scimind2460 committed:
Apply suggestions from code review

Co-authored-by: Trevor Gross <tg@trevorgross.com>

scimind2460 committed:
Refine complex numbers documentation and remove redundancy

Removed redundant statement about overhead from libgcc calls and clarified the purpose of the Complex type.

scimind2460 commented:
Hopefully I've cleared up everything to satisfaction. I suppose it's broadly reasonable to FCP?

kennytm commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,158 @@
+let a = first + second; // 4 + 6i
+let b = first - second; // -2 - 2i
+let c = first * second; // -5 + 10i
+let d = float_second / float_first; // 0.44 - 0.8imi
```
```suggestion
let d = float_second / float_first; // 0.44 - 0.8i
```

scimind2460 committed:
Update text/3892-complex-numbers.md

Co-authored-by: kennytm <kennytm@gmail.com>

tmandry commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,158 @@
+
+The multiple emitted calls to `libgcc.so` (`__mulsc3` and the like) via compiler-builtins may cause a bit of overhead and may not be what the Rust lang team and compiler team want.
+
+## Rationale and alternatives
```
I saw some reference to this in the discussion, but it's not addressed here: Why should this type, which is constrained to have the same layout, representation, and calling convention as C, go in `std::num` instead of `std::ffi`?

scimind2460 replied:
In addition to @programmerjake's great comments on complex numbers, @kennytm's [excellent comment here](https://github.com/rust-lang/rfcs/pull/3892?notification_referrer_id=NT_kwHOCtf34NoAJFJlcG9zaXRvcnk7MTc1Mjc5ODk7SXNzdWU7MzY4NTI2NTQxMQ#issuecomment-3615501813) sums up quite succinctly why we should not use polar numbers. However, to avoid future confusion, I will add it into the RFC.

scimind2460 committed:
Revise alternatives section for complex numbers

Clarify the rationale behind complex number implementation alternatives.

scimind2460 committed:
Enhance discussion on complex number alternatives

Expanded on the limitations of polar complex numbers and the need for a standard type for FFI with C. Discussed alternatives to complex number representation and their implications.

scimind2460 committed:
Expand alternatives section for complex numbers

Added a section discussing alternatives to complex number implementations, including polar layout and non-generic primitive types, while emphasizing the importance of FFI compatibility.

programmerjake commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,162 @@
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+`Complex<T>` numbers are in core::num and reexported in std::num, like `use core::num::Complex` or `use std::num::Complex`
+`Complex<T>` numbers can be instantiated with any component type using `Complex::new(re, im)` where `re` and `im` are of the same type ( includes all numbers).
```
IMO, except where it's completely generic over all types, it shouldn't need to work with `NonZero`/`Saturating`/`Wrapping`. That can be left for a later proposal (that may not even require a RFC).

scimind2460 committed:
Expand notes on complex number implementation

Add considerations for complex number features and compatibility.

scimind2460 committed:
Enhance discussion on Complex number features

Clarify the potential benefits of simplifying Complex number creation and discuss future support for Imaginary and Gaussian integers.

scimind2460 committed:
Refine discussion on complex number features

Clarify suggestions for complex number operations and future support.

tmandry commented on text/3892-complex-numbers.md:
```diff
@@ -0,0 +1,164 @@
+[reference-level-explanation]: #reference-level-explanation
+
+The `core` crate will provide implementations for operator traits for possible component types. For now, complex operations like `Mul` and `Div` are defined specifically for each floating-point type, while `Add` and `Sub` are defined for any components that themselves implement `Add` and `Sub`.
+Calls to some `libgcc` functions may also be needed, and will be emitted by the backend via compiler-builtins, specifically `__mulsc3`, `__muldc3`, `__divsc3` and `__divdc3` for the proper and complete implementation of these types. This is because if we implemented these operations generically, to implement any overflow checks, we would need to implement a trait that can generically call, say, a ...
```
This looks like an implementation-quality issue that might change, and is not part of the surface feature being approved by lang or libs-api. Assuming that's right, it does not belong in the reference-level section. I would recommend taking it out.

`libgcc` is obviously platform specific too.

> This is because if we implemented these operations generically, to implement any overflow checks, we would need to implement a trait that can generically call, say, a `max()` function and compare between values.

This doesn't seem all that complex to me, and the choice of whether to do this or rely on the compiler runtime falls within the set of tradeoffs normally made as implementation level decisions.

scimind2460 replied:
Unfortunately trait-level decisions were previously deemed as not implementable now by libs-api. I am new to the RFC process so if this is normally an implementation-level detail, I will remove it or put it in the Unresolved questions section.

scimind2460 committed:
Enhance RFC for complex numbers in Rust

Clarified the implementation details of complex number operations and their calling conventions with C. Added explanations for arithmetic operations and discussed the drawbacks of alternative representations.

scimind2460 commented:
@tmandry I hope that all issues are sufficiently resolved?

tmandry reviewed (approved):
@rfcbot reviewed

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3892/208151010f8cba1d69c354eb79dcbb6ee5b25dc4..5bbb20245a6a08476e8d136cf94465b4095e3fac)*

scimind2460 commented:
> I like the idea that i32 is 32 bits, u64 is 64 bits etc. Using c32 for a type with a size of 64 bits may be confusing or at least, error prone for users.

To be fair, one could also argue the other way, that it may be confusing for users who may think `c64` is actually `Complex<f64>` instead of `Complex<f32>` due to them sharing the same number. It really is a matter of perspective.

We could theoretically discuss which one is better here, but I don't think this should block the RFC.

Thanks for raising the point @lokyhark!

scimind2460 commented:
> If one needs a discussion to determine whether `c64` means `Complex<f64>` or `Complex<f32>` perhaps it's the best _not_ to introduce such type alias to avoid ambiguity at all.

Yes, I agree. It would essentially be clear looking at the generic type.

 Besides, we should have this discussion later, given that it's mentioned in the RFC as a future possibility.

traviscross committed:
Clean up whitespace, etc.

traviscross committed:
Add tracking issue for RC 3892
